File: solutions/cpp/nucleotide-count/1/nucleotide_count.cpp
#include <utility>

#include "nucleotide_count.h"

using namespace std;

namespace nucleotide_count
{

    counter::counter(string strand)
    {
        this->strand = strand;

        for (char c : strand)
        {
            switch (c)
            {
            case 'A':
            case 'C':
            case 'G':
            case 'T':
                counts.at(c)++;
                break;
            default:
                throw std::invalid_argument{""};
            }
        }
    };

    const std::map<char, int> counter::nucleotide_counts() const
    {
        return counts;
    };

    int counter::count(char c) const
    {
        switch (c)
        {
        case 'A':
        case 'C':
        case 'G':
        case 'T':
            return counts.at(c);
            break;
        default:
            throw std::invalid_argument{""};
        }
    }
} // namespace nucleotide_count

File: solutions/javascript/nucleotide-count/1/nucleotide-count.js
export class NucleotideCounts {
  static parse(dna) {
    let nucleotides = dna.split("");

    if (nucleotides.some((e) => !/[ACGT]/.test(e)))
      throw new Error("Invalid nucleotide in strand");

    const counts = {
      A: 0,
      C: 0,
      G: 0,
      T: 0,
    };

    nucleotides.forEach((e) => counts[e]++);

    return `${counts["A"] || 0} ${counts["C"] || 0} ${counts["G"]} ${
      counts["T"]
    }`;
  }
}
